File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let len = nums.length;
    let res = new Array(len);
    let lpdt = 1;

    for (let i = 0; i < len; i++) {
        res[i] = lpdt;
        lpdt *= nums[i];
    }

    let rpdt = 1;
    for (let i = len - 1; i >= 0; i--) {
        res[i] *= rpdt;
        rpdt *= nums[i];
    }

    return res;
};

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] r = new int[n];
        int[] ans = new int[n];
        int l = 1;
        int p = 1;

        for (int i = n - 1; i >= 0; i--) {
            p *= nums[i];
            r[i] = p;
        }
        for (int i = 0; i < n - 1; i++) {
            ans[i] = l * r[i + 1];
            l *= nums[i];
        }
        ans[n - 1] = l;
        return ans;
    }
}